### PR TITLE
Add support for Slots

### PR DESCRIPTION
**Pull request by victorhmp**
#### What does this PR do?

These changes aim to enable `vtex.render-runtime` to support the use of `slots` by Store Framework users.

The idea and motivation for Slots in Store Framework can be found at [Slots RFC](https://github.com/vtex-apps/store-discussion/issues/213)

#### How to test it?

In this [workspace](https://slotsplayground--storecomponents.myvtex.com/), there is a version of `vtex.store-theme` linked which does **not** use `blocks`, only `children` and Slots to pass components around. This is only possible due to this version of `render-runtime` and [this](https://github.com/vtex/builder-hub/pull/989) version of `builder-hub`.

Apps that receive Slots are also modified and linked in this workspace since they now receive components via **props** instead of just rendering `<ExtensionPoint />`s.

The full list of linked apps in this workspace is the following

Screen Shot 2020-04-14 at 15 42 01

To test for regressions, these accounts were used:

- [AdmitOne](https://victormiranda--admitone.myvtex.com/)
- [World-Wide Golf](https://victormiranda--worldwidegolf.myvtex.com/)
- [Texas Boot](https://victormiranda--txboot.myvtex.com/)
- [TBB](https://victormiranda--tbb.myvtex.com/)
- [Store Components](https://victormiranda--storecomponents.myvtex.com/)
- [Als](https://victormiranda--alssports.myvtex.com/)
- [Biscoind](https://victormiranda--biscoind.myvtex.com/)
- [Boticario](https://victormiranda--boticario.myvtex.com/)
- [Boticario Global](https://victormiranda--btglobal.myvtex.com/)
- [Exito Colombia](https://victormiranda--exitocol.myvtex.com/)
- [Madine Jeans](https://victormiranda--gc-jqu0734.myvtex.com/)
- [Garmin](https://victormiranda--garmin.myvtex.com/)
- [Samsung Arg](https://victormiranda--samsungar.myvtex.com/)
- [Ecowater](https://victormiranda--ecowaterqa.myvtex.com/)
- [Eriks Bike Shop](https://victormiranda--eriksbikeshop.myvtex.com/)
- [Red Oxx](https://victormiranda--redoxx.myvtex.com/)
- [Power Planet](https://victormiranda--powerplanet.myvtex.com/)
- [PagueMenos](https://victormiranda--paguemenos.myvtex.com/)
- [Motorola UK](https://victormiranda--motorolauk.myvtex.com/)
- [Melimelo Paris](https://victormiranda--melimeloparis.myvtex.com/)
- [Lions Pride](https://victormiranda--lionspride.myvtex.com/)

In all of these accounts, this version of `render-runtime` was installed in isolation, with no changes to the currently installed themes.

The regression test consisted of navigating through a few products, search, and custom pages to detect any anomalous behavior, such as missing components and unnecessary component remounts.

#### Describe alternatives you've considered if any.

This section is better explored in the RFC linked above.

#### Related to / Depends on \*

Depends on: https://github.com/vtex/builder-hub/pull/989

### CODE REVIEW

**Comment by victorhmp** [comment]
This is in draft due to the "flashing" components being observed during navigation. But I guess the base for these changes should not change much.

File: react/components/ExtensionPoint/ComponentLoader.tsx [hunk 1]
```diff
@@ -66,10 +66,16 @@ const ComponentLoader: FunctionComponent<Props> = props => {
     treePath,
     props: componentProps,
     hydration,
-    runtime,
   } = props
 
-  // Each prop that receives a slot should always be PascalCased.
+  /**
+   * Props that function as Slots are ALWAYS CamelCased.
```
Comment (Klynger) [inline react/components/ExtensionPoint/ComponentLoader.tsx:72]: PascalCased

File: react/components/ExtensionPoint/ComponentLoader.tsx [hunk 2]
```diff
@@ -66,10 +66,16 @@ const ComponentLoader: FunctionComponent<Props> = props => {
...
+   * It is OK to not include componentProps in the dependency array
```
Comment (Klynger) [inline react/components/ExtensionPoint/ComponentLoader.tsx:73]: Coudn't this be done by something above this component?

File: react/components/ExtensionPoint/index.tsx [hunk 3]
```diff
@@ -40,32 +40,38 @@ function mountTreePath(currentId: string, parentTreePath: string) {
...
+        block.blockRole === 'children') &&
+      block.blockRole !== 'slot'
+  )
+
+  return childBlocks.map((child, i) => {
+    const childTreePath = mountTreePath(child.extensionPointId, treePath)
+
+    const childExtension =
+      runtime.extensions && runtime.extensions[childTreePath]
+    const childProps = childExtension ? childExtension.props : {}
```
Comment (kaisermann) [inline react/components/ExtensionPoint/index.tsx:64]:
```suggestion
    const childProps = childExtension?.props ?? {}
```

File: react/utils/slots.tsx [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      treePath: treePath.replace(runtime.page, '*'),
+      pageContext: runtime.route.pageContext,
+    },
+  })
+
+  let extensionContent =
+    (data?.listContent[1]?.contentJSON &&
+      (JSON.parse(data.listContent[1]?.contentJSON) as Record<string, any>)) ??
+    (data?.listContent[0]?.contentJSON &&
+      (JSON.parse(data.listContent[0]?.contentJSON) as Record<string, any>))
```
Comment (kaisermann) [inline react/utils/slots.tsx:128]: wat

Comment (kaisermann) [reply react/utils/slots.tsx:128]: Woould it be possible to assign some of these values to variable? This is kind of cryptic to people from outside

File: react/components/ExtensionPoint/ComponentLoader.tsx (hunk 2 above)
Comment (victorhmp) [reply react/components/ExtensionPoint/ComponentLoader.tsx:73]: I couldn't make it work :(
Tried to use `useMemo` on `generateSlots` function without any success.

File: react/components/ExtensionPoint/ComponentLoader.tsx [hunk 5]
```diff
@@ -60,16 +68,65 @@ const ComponentLoader: FunctionComponent<Props> = props => {
...
+        (resultingSlotsProps[slotName] = generateSlot({
+          treePath,
+          slotName,
+          slotValue: componentProps[slotName],
+          hydration,
+        }))
+    )
+
+    return resultingSlotsProps
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (victorhmp) [reply react/components/ExtensionPoint/ComponentLoader.tsx:100]: It is explained in the comment above the `useMemo` call: https://github.com/vtex-apps/render-runtime/pull/510/files/cbaed71f1063a14a06d752927db8a3f8d26beb31#diff-96afb79b5ea04a051db365087912802fR73-R78

File: react/utils/slots.tsx (hunk 4 above)
Comment (victorhmp) [reply react/utils/slots.tsx:128]: Sure! I agree 😂

File: react/components/ExtensionPoint/ComponentLoader.tsx [hunk 6]
```diff
@@ -60,16 +68,65 @@ const ComponentLoader: FunctionComponent<Props> = props => {
...
+
+    slotNames.forEach(
+      slotName =>
+        (resultingSlotsProps[slotName] = generateSlot({
+          treePath,
+          slotName,
+          slotValue: componentProps[slotName],
+          hydration,
+        }))
+    )
```
Comment (kaisermann) [inline react/components/ExtensionPoint/ComponentLoader.tsx:97]: If you're not going to use a `reduce`, a `for-of` can be a tiny bit more readable than a `forEach`
```suggestion
const resultingSlotsProps: Record<string, ReactNode> = {}

for (const slotName of slotNames) {
  resultingSlotsProps[slotName] = generateSlot({
    treePath,
    slotName,
    slotValue: componentProps[slotName],
    hydration,
  })
}
```

File: react/components/ExtensionPoint/index.tsx [hunk 7]
```diff
@@ -40,32 +40,37 @@ function mountTreePath(currentId: string, parentTreePath: string) {
...
-        />
-      )
-    })
+  const childBlocks = extension.blocks.filter(
+    block =>
+      (block.children === undefined ||
+        block.children === true ||
+        block.blockRole === 'children') &&
+      block.blockRole !== 'slot'
+  )
```
Comment (kaisermann) [inline react/components/ExtensionPoint/index.tsx:57]:
```suggestion
const childBlocks = extension.blocks.filter(block => {
  const hasChildren = block.children === undefined || block.children === true || block.blockRole === 'children'
  const isNotSlot = block.blockRole !== 'slot'

  return hasChildren && isNotSlot
})
```

File: react/components/ExtensionPoint/ComponentLoader.tsx [hunk 8]
```diff
@@ -60,16 +68,64 @@ const ComponentLoader: FunctionComponent<Props> = props => {
...
+    () => ({
+      ...props,
+      props: componentPropsWithSlots,
+    }),
+    [componentPropsWithSlots, props]
+  )
+
+  const Component = useMemo(() => component && getImplementation(component), [
+    component,
+  ])
```
Comment (klzns) [inline react/components/ExtensionPoint/ComponentLoader.tsx:119]: `getImplementation` is just an access to an object this useMemo is kind of useless, or am I missing something?

```suggestion
  const Component = component && getImplementation(component)
```

File: react/typings/global.d.ts [hunk 9]
```diff
@@ -65,6 +65,7 @@ declare global {
     extensionPointId: string
     blockId: string
     children?: boolean
+    blockRole?: 'block' | 'children' | 'slot'
```
Comment (klzns) [inline react/typings/global.d.ts:68]: This property superseed the `children` property? If so, would be nice to add a annotation above `children`, what do you think?

File: react/utils/slots.tsx [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  const newTreePath = `${treePathWithoutSlotMarker}/${slotValue}`
+
+  const SlotComponent: FC<any> = memo(props => {
+    const runtime = useRuntime()
+    let extension = runtime.extensions[newTreePath]
+
+    const slotChildren = getChildExtensions(runtime, newTreePath)
+    let componentProps = extension?.props ?? {}
+
+    if (props.id) {
```
Comment (klzns) [inline react/utils/slots.tsx:51]: If it have an id, what does it mean? Mind adding comments?

File: react/utils/slots.tsx [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+      const hasLabel = slotValue.includes('#')
+      const dynamicTreePath = `${newTreePath}${hasLabel ? '-' : '#'}slot${
+        props.id
+      }`
+
+      if (!(dynamicTreePath in runtime.extensions)) {
```
Comment (klzns) [inline react/utils/slots.tsx:57]: What does that mean? Mind adding comments too?

Another way to comment the code is to assign this condition to a variable with a good name:

```ts
const slotHasNoSettings = !(dynamicTreePath in runtime.extensions)
if (slotHasNoSettings) {
```

File: react/utils/slots.tsx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+            {slotChildren}
+          </DynamicSlot>
+        )
+      }
+
+      extension = runtime.extensions[dynamicTreePath]
+      componentProps = extension?.props ?? {}
+    }
+
+    const extensionContent = extension?.content
```
Comment (klzns) [inline react/utils/slots.tsx:81]: Last but not least, a comment here too for the happy path

File: react/utils/slots.tsx [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+      template: runtime.page,
+      blockId,
+      treePath: treePath.replace(runtime.page, '*'),
+      pageContext: runtime.route.pageContext,
+    },
+  })
+
+  const contentFromCMS =
+    data?.listContent[1]?.contentJSON &&
+    (JSON.parse(data.listContent[1]?.contentJSON) as Record<string, any>)
```
Comment (klzns) [inline react/utils/slots.tsx:128]: Add useMemo here, since JSON.parse it not that cheap

File: react/utils/slots.tsx [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+  /**
+   * This default content for a certain blockId is often just the same values
+   * passed by users in their themes.
+   */
+  const defaultContentForBlockId =
+    data?.listContent[0]?.contentJSON &&
+    (JSON.parse(data.listContent[0]?.contentJSON) as Record<string, any>)
```
Comment (klzns) [inline react/utils/slots.tsx:135]: Same

File: react/utils/slots.tsx [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+  let extensionContentFromCMS = contentFromCMS ?? defaultContentForBlockId
+
+  const componentLoaderPropsWithContent = useMemo(
+    () =>
+      extensionContentFromCMS
+        ? { ...props, ...extensionContentFromCMS }
+        : props,
+    [extensionContentFromCMS, props]
```
Comment (klzns) [inline react/utils/slots.tsx:144]: _Maybe_ do everything inside of this?

File: react/utils/slots.tsx [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+  )
+
+  if (loading) {
+    return null
+  }
+  if (error) {
```
Comment (klzns) [inline react/utils/slots.tsx:150]:
```suggestion

  if (error) {
```

File: react/utils/slots.tsx [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+    runtime.extensions[treePath] = runtime.extensions[baseTreePath]
+    extensionContentFromCMS = runtime.extensions[baseTreePath]?.content
+  } else {
+    runtime.extensions[treePath] = runtime.extensions[baseTreePath]
+      ? {
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          ...runtime.extensions[baseTreePath]!,
+          content: extensionContentFromCMS,
+        }
+      : null
```
Comment (klzns) [inline react/utils/slots.tsx:165]: Mutating the `__RUNTIME__.extensions`, is that expected?

File: react/components/ExtensionPoint/ComponentLoader.tsx (hunk 8 above)
Comment (victorhmp) [reply react/components/ExtensionPoint/ComponentLoader.tsx:119]: Yep, you're right! I added this while trying to debug the unnecessary re-renders on this component, but I guess it is not really necessary :)

File: react/typings/global.d.ts (hunk 9 above)
Comment (victorhmp) [reply react/typings/global.d.ts:68]: Kind of... I cannot remove the `children` property from the build result generated from `buider-hub`, so this property will never be removed.

Comment (victorhmp) [reply react/typings/global.d.ts:74]: I've just added a comment pointing that out :)

**Pull request merged**